File: src/main/java/me/soknight/minigram/chats/model/dto/mapper/ChatDtoMapper.java
package me.soknight.minigram.chats.model.dto.mapper;

import lombok.AllArgsConstructor;
import me.soknight.minigram.chats.exception.ApiException;
import me.soknight.minigram.chats.model.dto.ChatDto;
import me.soknight.minigram.chats.model.dto.ChatMemberDto;
import me.soknight.minigram.chats.model.dto.ChatMessageDto;
import me.soknight.minigram.chats.model.entity.ChatEntity;
import me.soknight.minigram.chats.model.entity.ChatMemberEntity;
import me.soknight.minigram.chats.model.entity.ChatMessageEntity;
import me.soknight.minigram.chats.service.client.ProfileClient;
import me.soknight.minigram.chats.service.client.model.dto.ProfileDto;
import org.jspecify.annotations.NonNull;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Service
@AllArgsConstructor
public class ChatDtoMapper {

    private final @NonNull ProfileClient profileClient;

    public @NonNull ChatMemberDto toChatMemberDto(@NonNull ChatMemberEntity member) throws ApiException {
        return toChatMemberDto(member, new HashMap<>());
    }

    public @NonNull ChatDto toChatDto(@NonNull ChatEntity chat) throws ApiException {
        try {
            return toChatDto(chat, new HashMap<>());
        } catch (ChatDtoMappingException ex) {
            throw ex.apiException();
        }
    }

    public @NonNull ChatMessageDto toChatMessageDto(@NonNull ChatMessageEntity message) throws ApiException {
        try {
            var profiles = new HashMap<UUID, ProfileDto>();

            return new ChatMessageDto(
                    message.getMessageId(),
                    toChatDto(message.getChat(), profiles),
                    toChatMemberDto(message.getSender(), profiles),
                    message.getContent(),
                    message.getCreatedAt(),
                    message.getUpdatedAt()
            );
        } catch (ChatDtoMappingException ex) {
            throw ex.apiException();
        }
    }

    private @NonNull ChatDto toChatDto(
            @NonNull ChatEntity chat,
            @NonNull Map<UUID, ProfileDto> profiles
    ) throws ApiException {
        var members = chat.getMembers().stream()
                .sorted(Comparator.comparing(ChatMemberEntity::getJoinedAt).thenComparing(ChatMemberEntity::getUserId))
                .map(member -> toUncheckedChatMemberDto(member, profiles))
                .toList();

        return new ChatDto(
                chat.getId(),
                chat.getType(),
                chat.getTitle(),
                chat.getOwnerId(),
                members,
                chat.getLastMessageId(),
                chat.getCreatedAt(),
                chat.getUpdatedAt()
        );
    }

    private @NonNull ChatMemberDto toChatMemberDto(
            @NonNull ChatMemberEntity member,
            @NonNull Map<UUID, ProfileDto> profiles
    ) throws ApiException {
        var profile = getProfile(member.getUserId(), profiles);

        return new ChatMemberDto(
                member.getUserId(),
                profile.name(),
                profile.photoUrl(),
                member.getRole(),
                member.getJoinedAt()
        );
    }

    private @NonNull ChatMemberDto toUncheckedChatMemberDto(
            @NonNull ChatMemberEntity member,
            @NonNull Map<UUID, ProfileDto> profiles
    ) {
        try {
            return toChatMemberDto(member, profiles);
        } catch (ApiException ex) {
            throw new ChatDtoMappingException(ex);
        }
    }

    private @NonNull ProfileDto getProfile(
            UUID userId,
            @NonNull Map<UUID, ProfileDto> profiles
    ) throws ApiException {
        var profile = profiles.get(userId);
        if (profile != null) return profile;

        profile = profileClient.getProfile(userId);
        profiles.put(userId, profile);
        return profile;
    }

    public static final class ChatDtoMappingException extends RuntimeException {

        private final @NonNull ApiException apiException;

        public ChatDtoMappingException(@NonNull ApiException apiException) {
            super(apiException);
            this.apiException = apiException;
        }

        public @NonNull ApiException apiException() {
            return apiException;
        }

    }

}

File: src/main/kotlin/me/soknight/minigram/chats/service/ChatDtoMapper.kt
@file:OptIn(ExperimentalUuidApi::class)

package me.soknight.minigram.chats.service

import me.soknight.minigram.chats.client.ProfileClient
import me.soknight.minigram.chats.domain.ChatMemberRow
import me.soknight.minigram.chats.domain.ChatMessageRow
import me.soknight.minigram.chats.domain.ChatRow
import me.soknight.minigram.chats.dto.ChatDto
import me.soknight.minigram.chats.dto.ChatMemberDto
import me.soknight.minigram.chats.dto.ChatMessageDto
import me.soknight.minigram.chats.dto.ProfileDto
import me.soknight.minigram.chats.repository.ChatMemberRepository
import me.soknight.minigram.chats.repository.ChatRepository
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid
import org.koin.core.annotation.Single

@Single
class ChatDtoMapper(
    private val chatRepository: ChatRepository,
    private val memberRepository: ChatMemberRepository
) {
    suspend fun toChatDto(chat: ChatRow, profileClient: ProfileClient): ChatDto {
        val members = memberRepository.findByChatId(chat.id, 0, 200)
        val cache   = mutableMapOf<Uuid, ProfileDto>()
        return ChatDto(
            id            = chat.id,
            type          = chat.type,
            title         = chat.title,
            ownerId       = chat.ownerId,
            members       = members.map { toChatMemberDto(it, profileClient, cache) },
            lastMessageId = chat.lastMessageId,
            createdAt     = chat.createdAt,
            updatedAt     = chat.updatedAt
        )
    }

    suspend fun toChatMemberDto(
        member: ChatMemberRow,
        profileClient: ProfileClient,
        cache: MutableMap<Uuid, ProfileDto> = mutableMapOf()
    ): ChatMemberDto {
        val profile = cache.getOrPut(member.userId) { profileClient.getProfile(member.userId) }
        return ChatMemberDto(member.userId, profile.name, profile.photoUrl, member.role, member.joinedAt)
    }

    suspend fun toChatMessageDto(message: ChatMessageRow, profileClient: ProfileClient): ChatMessageDto {
        val chat   = chatRepository.findById(message.chatId) ?: error("Chat ${message.chatId} not found")
        val cache  = mutableMapOf<Uuid, ProfileDto>()
        val chatDto = toChatDto(chat, profileClient)
        val senderRow = memberRepository.findById(message.chatId, message.senderId)
            ?: error("Sender ${message.senderId} not found in chat ${message.chatId}")
        return ChatMessageDto(
            id        = message.messageId,
            chat      = chatDto,
            sender    = toChatMemberDto(senderRow, profileClient, cache),
            content   = message.content,
            createdAt = message.createdAt,
            updatedAt = message.updatedAt
        )
    }
}
